File: Level2/[스택]올바른 괄호/Solution.java
import java.util.*;

class Solution {
    public static void main(String[] args) {
        String s = "()()";
        String s2 = "(())()";
        String s3 = ")()(";
        String s4 = "(()(";
        if(solution(s))
            System.out.println("True");
        else
            System.out.println("false");
    }

    public static boolean solution(String s) {
        boolean answer = true;
        Stack<Character> stack = new Stack<>();
        char[] chs = s.toCharArray();
        
        for(int i=0, size=chs.length; i<size; i++){
            switch(chs[i]){
                case '(':
                    stack.push(chs[i]);
                    break;
                case ')':
                    if(stack.size() != 0)
                        stack.pop();
                    else
                        return false;
                    break;
            }
        }
        
        if(stack.size() != 0)
            return false;
        return answer;
    }
}

File: Level2/[스택]올바른 괄호/solution.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

bool solution(string s) {
	stack<char> st;

	for (int i = 0; i < s.length(); i++) {
		if (s.at(i) == '(') {
			st.push('(');
		}
		else {
			if (st.size() != 0) {
				st.pop();
			}
			else {
				return false;
			}
		}
	}

	if (st.size() != 0) {
		return false;
	}
	return true;
}

int main() {
	// IO 속도 향상
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	string s = "(()(";

	cout << solution(s) << "\n";

	return 0;
}